base_layer/core/src/consensus/consensus_manager.rs

```diff
@@ -20,36 +20,144 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-use crate::consensus::emission::EmissionSchedule;
+use crate::{
+    chain_storage::{BlockchainBackend, ChainStorageError},
+    consensus::emission::EmissionSchedule,
+    proof_of_work::{DiffAdjManager, DiffAdjManagerError, Difficulty, DifficultyAdjustmentError, PowAlgorithm},
+};
+use derive_error::Error;
+use std::sync::{Arc, RwLock};
+use tari_utilities::epoch_time::EpochTime;
 
-/// This is the used to control all consensus values.
-pub struct ConsensusManager {
-    /// The emission schedule to use for coinbase rewards
-    emission_schedule: EmissionSchedule,
+#[derive(Debug, Error, Clone, PartialEq)]
+pub enum ConsensusManagerError {
+    /// Difficulty adjustment encountered an error
+    DifficultyAdjustmentError(DifficultyAdjustmentError),
+    /// Difficulty adjustment manager encountered an error
+    DifficultyAdjustmentManagerError(DiffAdjManagerError),
+    /// Problem with the DB backend storage
+    ChainStorageError(ChainStorageError),
+    /// There is no blockchain to query
+    EmptyBlockchain,
+    /// RwLock access broken
+    PoisonedAccess,
```

Comment:
I always prefer mapping this to some other error with a string message. It's very much a leaky abstraction that you can't really handle from outside the struct with the RWLock